File: pdf-generator-service/src/main/java/br/com/bra/pdfgeneratorservice/kafka/common/KafkaProducerSettings.java
package br.com.bra.pdfgeneratorservice.kafka.common;


import io.confluent.kafka.serializers.KafkaAvroSerializer;
import io.confluent.kafka.serializers.KafkaAvroSerializerConfig;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringSerializer;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.core.DefaultKafkaProducerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.core.ProducerFactory;

import java.util.HashMap;
import java.util.Map;

@Configuration
public class KafkaProducerSettings {
    @Value("${kafka-config.server-url}")
    private String serverUrl;
    @Value("${kafka-config.schema-registry-url}")
    private String schemaRegistryUrl;

    @Bean
    KafkaTemplate<String, Object> kafkaTemplate() {
        return new KafkaTemplate<>(producerFactory());
    }

    private ProducerFactory<String, Object> producerFactory() {
        Map<String, Object> props = new HashMap<>();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, serverUrl);
        props.put(ProducerConfig.ACKS_CONFIG, "all");
        props.put(ProducerConfig.LINGER_MS_CONFIG, 10);
        props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, false);
        props.put(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, 2);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, KafkaAvroSerializer.class);
        props.put(KafkaAvroSerializerConfig.SCHEMA_REGISTRY_URL_CONFIG, schemaRegistryUrl);
        props.put(KafkaAvroSerializerConfig.AUTO_REGISTER_SCHEMAS, true);

        return new DefaultKafkaProducerFactory<>(props);
    }
}

File: processing-service/src/main/kotlin/br/com/bra/processingservice/kafka/settings/KafkaProducerSettings.kt
package br.com.bra.processingservice.kafka.settings

import io.confluent.kafka.serializers.KafkaAvroSerializer
import io.confluent.kafka.serializers.KafkaAvroSerializerConfig
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.serialization.StringSerializer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.core.DefaultKafkaProducerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.core.ProducerFactory
import java.util.*

@Configuration
class KafkaProducerSettings(
    private val kafkaProperties: LocalKafkaProperties
) {
    @Bean
    fun kafkaTemplate(): KafkaTemplate<String, Objects> = KafkaTemplate(producerFactory())
    private fun producerFactory(): ProducerFactory<String, Objects> = hashMapOf<String, Any>()
        .apply {
            this[ProducerConfig.BOOTSTRAP_SERVERS_CONFIG] = kafkaProperties.serverUrl
            this[ProducerConfig.ACKS_CONFIG] = "all"
            this[ProducerConfig.LINGER_MS_CONFIG] = 10
            this[ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG] = false
            this[ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION] = 2
            this[ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG] = StringSerializer::class.java
            this[ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG] = KafkaAvroSerializer::class.java
            this[KafkaAvroSerializerConfig.SCHEMA_REGISTRY_URL_CONFIG] = kafkaProperties.schemaRegistryUrl
            this[KafkaAvroSerializerConfig.AUTO_REGISTER_SCHEMAS] = true
        }.let { DefaultKafkaProducerFactory(it) }
}